src/android_utils.py

```diff
@@ -107,6 +109,52 @@ def get_home_folder():
     return os.path.join(kolibri_home_file.toString(), "KOLIBRI_DATA")
 
 
+def get_endless_key_paths():
+    for external_file_dir in get_activity().getExternalFilesDirs(None):
+        state = Environment.getExternalStorageState(external_file_dir)
+        is_removable = Environment.isExternalStorageRemovable(external_file_dir)
+        if is_removable and state == "mounted":
+            # Leave only the /storage/STORAGE_ID/
+            external_drive = Path(external_file_dir.toString()).parents[3]
+            kolibri_data_path = os.path.join(external_drive, "KOLIBRI_DATA")
+            content_path = os.path.join(kolibri_data_path, "content")
+            db_path = os.path.join(
+                kolibri_data_path, "preseeded_kolibri_home", "db.sqlite3"
+            )
+            if os.path.exists(content_path) and os.path.exists(db_path):
+                return {"content_path": content_path, "db_path": db_path}
+    return None
+
+
+def provision_endless_key_database(endless_key_paths):
+    if endless_key_paths is not None:
+        home_folder = get_home_folder()
+        if os.path.exists(os.path.join(home_folder, "db.sqlite3")):
+            logger.debug("EK database already exists, skipping.")
+            return
+        if not os.path.exists(home_folder):
+            os.mkdir(home_folder)
+        if Environment.isExternalStorageManager():
+            shutil.copy(endless_key_paths["db_path"], home_folder)
+            logger.debug("EK database provisioned.")
+        else:
+            logger.debug("EK database found in external storage bu user didn't allow.")
+
+
+def ask_all_files_access():
+    if SDK_INT < 30:
+        return
+    askIntent = Intent(
+        Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION,
+        Uri.parse("package:org.endlessos.Key"),
+    )
+    get_activity().startActivity(askIntent)
+
+
+def has_all_files_access():
+    return Environment.isExternalStorageManager()
```

Comment:
On my device (Chromebook Duet, ChromeOS 100, Android 9), this throws an exception…

```
05-20 12:45:15.558  5243  5325 I python  :   File "/home/jenkins/workspace/kolibri-installer-android-pr/src/main.py", line 67, in <module>
05-20 12:45:15.559  5243  5325 I python  :   File "/home/jenkins/workspace/kolibri-installer-android-pr/src/android_utils.py", line 155, in has_all_files_access
05-20 12:45:15.559  5243  5325 I python  : AttributeError: type object 'android.os.Environment' has no attribute 'isExternalStorageManager'
```

I see the device you're using has Android 11 now. I wonder if we need to do some feature probing?

Reply:
Ah good point! Yes I was on Android 9 before and then I updated this PR to make it work for Android 11.